qutip/core/data/expect.pyx

```diff
@@ -21,19 +21,27 @@ __all__ = [
 ]
 
 cdef void _check_shape_ket(Data op, Data state) nogil except *:
-    if op.shape[1] != state.shape[0] or state.shape[1] != 1:
+    if (op.shape[1] != state.shape[0]  # Matrix multiplication
+        or state.shape[1] != 1  # State is ket
+        or op.shape[0] != op.shape[1]  # op must be square matrix
+       ):
         raise ValueError("incorrect input shapes "
                          + str(op.shape) + " and " + str(state.shape))
 
 cdef void _check_shape_dm(Data op, Data state) nogil except *:
-    if op.shape[0] != state.shape[1] or op.shape[1] != state.shape[0]:
+    if (op.shape[0] != op.shape[1]  # Matrix multiplication
+        or state.shape[0] != state.shape[1]  # State is square 
+        or op.shape[1] != state.shape[0]  # Op is square
```

Comment:
```suggestion
    if (op.shape[1] != state.shape[0]  # Matrix multiplication
        or state.shape[0] != state.shape[1]  # State is square 
        or op.shape[0] != op.shape[1]  # Op is square
```
Fix ordering of checks to match comments and to be consistent with the comments and with the ordering of the ket checks above.